parsil/src/assembler.rs

```diff
@@ -388,25 +419,27 @@ impl<'a, C: ContextProvider> Assembler<'a, C> {
                         };
                         let bound = Some(self.expression_to_boundary(right)?);
 
-                        if self.secondary_index_bounds.high.is_some() {
-                            // impossible to say which is lower between two
-                            // conflicting high bounds
-                            self.secondary_index_bounds.high = None;
+                        if bounds.high.is_some() {
+                            match bound.partial_cmp(&bounds.high) {
+                                Some(Ordering::Less) => bounds.high = bound,
+                                // impossible to say which is lower between two
+                                // conflicting high bounds
+                                Some(_) => {}
+                                None => bounds.low = None,
+                            }
                         } else {
-                            self.secondary_index_bounds.high = bound;
+                            bounds.high = bound;
                         }
                     }
                     // $sid = x
                     BinaryOperator::Eq => {
                         let bound = Some(self.expression_to_boundary(right)?);
-                        if self.secondary_index_bounds.low.is_some()
-                            && self.secondary_index_bounds.high.is_some()
-                        {
-                            self.secondary_index_bounds.low = None;
-                            self.secondary_index_bounds.high = None;
+                        if bounds.low.is_some() && bounds.high.is_some() {
```

Comment:
If there is an equality predicate on secondary index which is in `AND` with all other predicates of the query, then isn't it always safe to use this as the query bounds? Like if there is a query `SELECT * FROM t WHERE sec_index BETWEEN $1 AND 45 AND sec_index = 34`, it should be ok to set secondary index bounds to `34` (as any matching row need to satisfy `sec_index = 34`, and the other predicate `sec_index BETWEEN $1 AND 45` will still be evaluated on the rows with `sec_index = 34`). Wdyt?

Reply:
So basically, it means that it's better to have a potentially non-optimal bound than no bound at all? Yes, that makes perfect sense to me – not sure why we didn't do that earlier.